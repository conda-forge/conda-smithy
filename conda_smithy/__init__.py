try:
    from conda_smithy._version import __version__
except ImportError:
    __version__ = "0.0.0"
