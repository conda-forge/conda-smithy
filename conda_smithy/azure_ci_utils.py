import os
import typing
import warnings

from msrest.authentication import Authentication, BasicAuthentication
from vsts.build.v4_1.build_client import BuildClient
from vsts.build.v4_1.models import (
    BuildDefinition,
    BuildDefinitionReference,
    SourceRepositories,
    SourceRepository,
)
from vsts.service_endpoint.v4_1.models import ServiceEndpoint
from vsts.service_endpoint.v4_1.service_endpoint_client import (
    ServiceEndpointClient,
)
from vsts.task_agent.v4_0.models import TaskAgentQueue
from vsts.task_agent.v4_0.task_agent_client import TaskAgentClient
from vsts.vss_connection import VssConnection


class AzureConfig:

    _default_org = "conda-forge"
    _default_project_name = "feedstock-builds"

    def __init__(
        self, org_or_user=None, project_name=None, team_instance=None
    ):
        self.org_or_user = org_or_user or os.getenv(
            "AZURE_ORG_OR_USER", self._default_org
        )
        # This will only need to be changed if you need to point to a non-standard azure
        # instance.
        self.instance_base_url = team_instance or os.getenv(
            "AZURE_INSTANCE", f"https://dev.azure.com/{self.org_or_user}"
        )
        self.project_name = project_name or os.getenv(
            "AZURE_PROJECT_NAME", self._default_project_name
        )

        try:
            with open(
                os.path.expanduser("~/.conda-smithy/azure.token"), "r"
            ) as fh:
                self.token = fh.read().strip()
            if not self.token:
                raise ValueError()
        except (IOError, ValueError):
            print(
                "No azure token. Create a token and\n"
                "put it in ~/.conda-smithy/azure.token"
            )
            self.token = None

        # By default for now don't report on the build information back to github
        self.azure_report_build_status = os.getenv(
            "AZURE_REPORT_BUILD_STATUS", "true"
        )

    @property
    def connection(self):
        connection = VssConnection(
            base_url=self.instance_base_url, creds=self.credentials
        )
        return connection

    @property
    def credentials(self):
        if self.token:
            return BasicAuthentication("", self.token)
        else:
            warnings.warn(
                "No token available.  No modifications will be possible!"
            )
            return Authentication()


default_config = AzureConfig()


def get_service_endpoint(config: AzureConfig = default_config):
    service_endpoint_client = ServiceEndpointClient(
        base_url=config.instance_base_url, creds=config.credentials
    )
    endpoints: typing.List[
        ServiceEndpoint
    ] = service_endpoint_client.get_service_endpoints(
        project=config.project_name, type="GitHub"
    )
    for service_endpoint in endpoints:
        if service_endpoint.name == config.org_or_user:
            return service_endpoint
    else:
        raise KeyError("Service endpoint not found")


def get_queues(
    config: AzureConfig = default_config,
) -> typing.List[TaskAgentQueue]:
    aclient = TaskAgentClient(config.instance_base_url, config.credentials)
    return aclient.get_agent_queues(config.project_name)


def get_default_queue(project_name):
    queues = get_queues(project_name)
    for q in queues:
        if q.name == "Default":
            return q
    else:
        raise ValueError("Default queue not found")


def get_repo_reference(config: AzureConfig, github_org, repo_name):
    service_endpoint = get_service_endpoint(config)
    bclient: BuildClient = config.connection.get_client(
        "vsts.build.v4_1.build_client.BuildClient"
    )
    repos: SourceRepositories = bclient.list_repositories(
        project=config.project_name,
        provider_name="github",
        repository=f"{github_org}/{repo_name}",
        service_endpoint_id=service_endpoint.id,
    )

    repo: SourceRepository = repos.repositories[0]
    return repo


def get_default_build_definition(
    github_org, repo_name, config: AzureConfig = default_config, **kwargs,
):
    from vsts.build.v4_1.models import (
        BuildDefinition,
        BuildRepository,
    )
    from vsts.task_agent.v4_0.task_agent_client import TaskAgentClient
    import inspect

    aclient = TaskAgentClient(config.instance_base_url, config.credentials)

    source_repo = get_repo_reference(config, github_org, repo_name)

    new_repo = BuildRepository(
        type="GitHub",
        url=source_repo.properties["cloneUrl"],
        **{
            k: v
            for k, v in source_repo.as_dict().items()
            if k in set(inspect.getfullargspec(BuildRepository).args) - {"url"}
        },
    )
    new_repo.name = source_repo.properties["fullName"]
    new_repo.properties["cleanOptions"] = "0"
    new_repo.properties["skipSyncSource"] = "false"
    new_repo.properties["gitLfsSupport"] = "false"
    new_repo.properties["checkoutNestedSubmodules"] = "false"
    new_repo.properties["labelSources"] = "0"
    new_repo.properties["fetchDepth"] = "0"
    new_repo.properties["labelSourcesFormat"] = "$(build.buildNumber)"
    new_repo.properties["reportBuildStatus"] = config.azure_report_build_status
    new_repo.clean = False

    queues = get_queues(config)
    default_queue = get_default_queue(config)
    service_endpoint = get_service_endpoint(config)

    build_definition = BuildDefinition(
        process={
            "type": 2,
            # These might be optional;
            "resources": {
                "queues": [{"id": q.id, "alias": q.name} for q in queues],
                "endpoints": [
                    {"id": service_endpoint.id, "alias": service_endpoint.name}
                ],
            },
            "yamlFilename": "/azure-pipelines.yml",
        },
        # queue works
        queue=default_queue,
        # now onto this
        repository=new_repo,
        name=repo_name,
        # configure trigger for our builds.
        triggers=[
            {
                "branchFilters": ["+*"],
                "forks": {"enabled": True, "allowSecrets": False},
                "pathFilters": [],
                "isCommentRequiredForPullRequest": False,
                "triggerType": "pullRequest",
            },
            {
                "branchFilters": ["+*"],
                "pathFilters": [],
                "batchChanges": False,
                "maxConcurrentBuildsPerBranch": 1,
                "pollingInterval": 0,
                "triggerType": "continuousIntegration",
            },
        ],
        variable_groups=aclient.get_variable_groups(
            project=config.project_name, group_name="anaconda-org"
        ),
        type="build",
        **kwargs,
    )

    return build_definition


def register_repo(github_org, repo_name, config: AzureConfig = default_config):
    from vsts.build.v4_1.models import BuildDefinitionReference

    bclient = build_client()
    build_definition = get_default_build_definition(
        github_org, repo_name, config=config,
    )

    # clean up existing builds for the same feedstock if present
    existing_definitions: typing.List[
        BuildDefinitionReference
    ] = bclient.get_definitions(project=config.project_name, name=repo_name)
    if existing_definitions:
        assert len(existing_definitions) == 1
        ed = existing_definitions[0]
        bclient.update_definition(
            definition=build_definition,
            definition_id=ed.id,
            project=ed.project.name,
        )
    else:
        bclient.create_definition(
            definition=build_definition, project=config.project_name
        )


def build_client(config: AzureConfig = default_config) -> BuildClient:
    return config.connection.get_client(
        "vsts.build.v4_1.build_client.BuildClient"
    )


def repo_registered(
    github_org: str, repo_name: str, config: AzureConfig = default_config
) -> bool:
    existing_definitions: typing.List[BuildDefinitionReference] = build_client(
        config
    ).get_definitions(project=config.project_name, name=repo_name)

    return bool(existing_definitions)


def enable_reporting(repo, config: AzureConfig = default_config) -> None:
    bclient = build_client(config)
    bdef_header = bclient.get_definitions(
        project=config.project_name, name=repo
    )[0]
    bdef = bclient.get_definition(bdef_header.id, bdef_header.project.name)
    bdef.repository.properties["reportBuildStatus"] = "true"
    bclient.update_definition(bdef, bdef.id, bdef.project.name)


def get_build_id(repo, config: AzureConfig = default_config) -> dict:
    """Get the necessary build information to persist in the config file to allow rendering
    of badges.
    This is needed by non-conda-forge use cases"""
    bclient = build_client(config)
    bdef_header = bclient.get_definitions(
        project=config.project_name, name=repo
    )[0]
    bdef: BuildDefinition = bclient.get_definition(
        bdef_header.id, bdef_header.project.name
    )

    return dict(
        user_or_org=config.org_or_user,
        project_name=config.project_name,
        build_id=bdef.id,
        project_id=bdef.project.id,
    )
