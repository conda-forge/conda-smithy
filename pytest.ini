[pytest]
markers =
    legacy_circle: Test designed to run as if prior to the azure migration
    legacy_travis: Test designed to run as if prior to the azure migration
    legacy_appveyor: Test designed to run as if prior to the azure migration
    cli: CLI tests outside of test/test_cli.py