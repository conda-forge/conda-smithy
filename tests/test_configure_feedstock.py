import os
import conda_smithy.configure_feedstock as cnfgr_fdstk

import pytest


def test_noarch_skips_appveyor(noarch_recipe, jinja_env):
    cnfgr_fdstk.render_appveyor(jinja_env=jinja_env,
                                forge_config=noarch_recipe.config,
                                forge_dir=noarch_recipe.recipe)
    # this configuration should be skipped
    assert not noarch_recipe.config['appveyor']['enabled']
    # no appveyor.yaml should have been written.  Nothing else, either, since we only ran
    #     appveyor render.  No matrix dir should exist.
    assert not os.path.isdir(os.path.join(noarch_recipe.recipe, '.ci_support'))


def test_noarch_skips_travis(noarch_recipe, jinja_env):
    cnfgr_fdstk.render_travis(jinja_env=jinja_env,
                              forge_config=noarch_recipe.config,
                              forge_dir=noarch_recipe.recipe)
    # this configuration should be skipped
    assert not noarch_recipe.config['travis']['enabled']
    # no appveyor.yaml should have been written.  Nothing else, either, since we only ran
    #     appveyor render.  No matrix dir should exist.
    assert not os.path.isdir(os.path.join(noarch_recipe.recipe, '.ci_support'))


def test_noarch_runs_on_circle(noarch_recipe, jinja_env):
    cnfgr_fdstk.render_circle(jinja_env=jinja_env,
                              forge_config=noarch_recipe.config,
                              forge_dir=noarch_recipe.recipe)
    # this configuration should be run
    assert noarch_recipe.config['circle']['enabled']
    # no appveyor.yaml should have been written.  Nothing else, either, since we only ran
    #     appveyor render.  No matrix dir should exist.
    matrix_dir = os.path.join(noarch_recipe.recipe, '.ci_support')
    assert os.path.isdir(matrix_dir)
    # single matrix entry - readme is generated later in main function
    assert len(os.listdir(matrix_dir)) == 1


def test_r_skips_appveyor(r_recipe, jinja_env):
    cnfgr_fdstk.render_appveyor(jinja_env=jinja_env,
                                forge_config=r_recipe.config,
                                forge_dir=r_recipe.recipe)
    # this configuration should be skipped
    assert not r_recipe.config['appveyor']['enabled']
    # no appveyor.yaml should have been written.  Nothing else, either, since we only ran
    #     appveyor render.  No matrix dir should exist.
    assert not os.path.isdir(os.path.join(r_recipe.recipe, '.ci_support'))


def test_r_matrix_travis(r_recipe, jinja_env):
    cnfgr_fdstk.render_travis(jinja_env=jinja_env,
                              forge_config=r_recipe.config,
                              forge_dir=r_recipe.recipe)
    # this configuration should be run
    assert r_recipe.config['travis']['enabled']
    # no appveyor.yaml should have been written.  Nothing else, either, since we only ran
    #     appveyor render.  No matrix dir should exist.
    matrix_dir = os.path.join(r_recipe.recipe, '.ci_support')
    assert os.path.isdir(matrix_dir)
    # single matrix entry - readme is generated later in main function
    assert len(os.listdir(matrix_dir)) == 2


def test_r_matrix_on_circle(r_recipe, jinja_env):
    cnfgr_fdstk.render_circle(jinja_env=jinja_env,
                              forge_config=r_recipe.config,
                              forge_dir=r_recipe.recipe)
    # this configuration should be run
    assert r_recipe.config['circle']['enabled']
    # no appveyor.yaml should have been written.  Nothing else, either, since we only ran
    #     appveyor render.  No matrix dir should exist.
    matrix_dir = os.path.join(r_recipe.recipe, '.ci_support')
    assert os.path.isdir(matrix_dir)
    # single matrix entry - readme is generated later in main function
    assert len(os.listdir(matrix_dir)) == 2


def test_py_matrix_appveyor(py_recipe, jinja_env):
    cnfgr_fdstk.render_appveyor(jinja_env=jinja_env,
                                forge_config=py_recipe.config,
                                forge_dir=py_recipe.recipe)
    # this configuration should be skipped
    assert py_recipe.config['appveyor']['enabled']
    matrix_dir = os.path.join(py_recipe.recipe, '.ci_support')
    assert os.path.isdir(matrix_dir)
    # 2 python versions, 2 target_platforms
    assert len(os.listdir(matrix_dir)) == 4


def test_py_matrix_travis(py_recipe, jinja_env):
    cnfgr_fdstk.render_travis(jinja_env=jinja_env,
                              forge_config=py_recipe.config,
                              forge_dir=py_recipe.recipe)
    # this configuration should be run
    assert py_recipe.config['travis']['enabled']
    # no appveyor.yaml should have been written.  Nothing else, either, since we only ran
    #     appveyor render.  No matrix dir should exist.
    matrix_dir = os.path.join(py_recipe.recipe, '.ci_support')
    assert os.path.isdir(matrix_dir)
    # single matrix entry - readme is generated later in main function
    assert len(os.listdir(matrix_dir)) == 2


def test_py_matrix_on_circle(py_recipe, jinja_env):
    cnfgr_fdstk.render_circle(jinja_env=jinja_env,
                              forge_config=py_recipe.config,
                              forge_dir=py_recipe.recipe)
    # this configuration should be run
    assert py_recipe.config['circle']['enabled']
    # no appveyor.yaml should have been written.  Nothing else, either, since we only ran
    #     appveyor render.  No matrix dir should exist.
    matrix_dir = os.path.join(py_recipe.recipe, '.ci_support')
    assert os.path.isdir(matrix_dir)
    # single matrix entry - readme is generated later in main function
    assert len(os.listdir(matrix_dir)) == 2


def test_circle_with_yum_reqs(py_recipe, jinja_env):
    with open(os.path.join(py_recipe.recipe, 'recipe', 'yum_requirements.txt'), 'w') as f:
        f.write('nano\n')
    cnfgr_fdstk.render_circle(jinja_env=jinja_env,
                              forge_config=py_recipe.config,
                              forge_dir=py_recipe.recipe)


def test_circle_with_empty_yum_reqs_raises(py_recipe, jinja_env):
    with open(os.path.join(py_recipe.recipe, 'recipe', 'yum_requirements.txt'), 'w') as f:
        f.write('# effectively empty')
    with pytest.raises(ValueError):
        cnfgr_fdstk.render_circle(jinja_env=jinja_env,
                                  forge_config=py_recipe.config,
                                  forge_dir=py_recipe.recipe)
