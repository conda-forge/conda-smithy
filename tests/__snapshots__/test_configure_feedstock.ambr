# serializer version: 1
# name: test_automerge_action_exists[conda-build]
  dict({
    'jobs': dict({
      'automerge-action': dict({
        'name': 'automerge',
        'runs-on': 'ubuntu-latest',
        'steps': list([
          dict({
            'id': 'automerge-action',
            'name': 'automerge-action',
            'uses': 'conda-forge/automerge-action@main',
            'with': dict({
              'github_token': '${{ secrets.GITHUB_TOKEN }}',
              'rerendering_github_token': '${{ secrets.RERENDERING_GITHUB_TOKEN }}',
            }),
          }),
        ]),
      }),
    }),
    True: dict({
      'check_suite': dict({
        'types': list([
          'completed',
        ]),
      }),
      'status': dict({
      }),
    }),
  })
# ---
# name: test_automerge_action_exists[rattler-build]
  dict({
    'jobs': dict({
      'automerge-action': dict({
        'name': 'automerge',
        'runs-on': 'ubuntu-latest',
        'steps': list([
          dict({
            'id': 'automerge-action',
            'name': 'automerge-action',
            'uses': 'conda-forge/automerge-action@main',
            'with': dict({
              'github_token': '${{ secrets.GITHUB_TOKEN }}',
              'rerendering_github_token': '${{ secrets.RERENDERING_GITHUB_TOKEN }}',
            }),
          }),
        ]),
      }),
    }),
    True: dict({
      'check_suite': dict({
        'types': list([
          'completed',
        ]),
      }),
      'status': dict({
      }),
    }),
  })
# ---
# name: test_choco_install[conda-build]
  '''
  # This file was generated automatically from conda-smithy. To update this configuration,
  # update the conda-forge.yml and/or the recipe/meta.yaml.
  # -*- mode: yaml -*-
  
  jobs:
  - job: win
    pool:
      vmImage: windows-2022
    strategy:
      matrix:
        win_64_c_compilervs2008python2.7:
          CONFIG: win_64_c_compilervs2008python2.7
          UPLOAD_PACKAGES: 'True'
        win_64_c_compilervs2015python3.5:
          CONFIG: win_64_c_compilervs2015python3.5
          UPLOAD_PACKAGES: 'True'
    timeoutInMinutes: 360
    variables:
      CONDA_BLD_PATH: D:\\bld\\
      UPLOAD_TEMP: D:\\tmp
  
    steps:
      - script: |
          choco install pkg0 -fdv -y --debug
        displayName: "Install Chocolatey Package: pkg0"
  
      - script: |
          choco install pkg1 --version=X.Y.Z -fdv -y --debug
        displayName: "Install Chocolatey Package: pkg1 --version=X.Y.Z"
  
  
      - task: PythonScript@0
        displayName: 'Download Miniforge'
        inputs:
          scriptSource: inline
          script: |
            import urllib.request
            url = 'https://github.com/conda-forge/miniforge/releases/latest/download/Mambaforge-Windows-x86_64.exe'
            path = r"$(Build.ArtifactStagingDirectory)/Miniforge.exe"
            urllib.request.urlretrieve(url, path)
  
      - script: |
          start /wait "" %BUILD_ARTIFACTSTAGINGDIRECTORY%\Miniforge.exe /InstallationType=JustMe /RegisterPython=0 /S /D=C:\Miniforge
        displayName: Install Miniforge
  
      - powershell: Write-Host "##vso[task.prependpath]C:\Miniforge\Scripts"
        displayName: Add conda to PATH
  
      - script: |
          call ".scripts\run_win_build.bat"
        displayName: Run Windows build
        env:
          PYTHONUNBUFFERED: 1
          CONFIG: $(CONFIG)
          CI: azure
          flow_run_id: azure_$(Build.BuildNumber).$(System.JobAttempt)
          remote_url: $(Build.Repository.Uri)
          sha: $(Build.SourceVersion)
          UPLOAD_PACKAGES: $(UPLOAD_PACKAGES)
          UPLOAD_TEMP: $(UPLOAD_TEMP)
          UPLOAD_ON_BRANCH: foo-branch
          BINSTAR_TOKEN: $(BINSTAR_TOKEN)
  '''
# ---
# name: test_choco_install[rattler-build]
  '''
  # This file was generated automatically from conda-smithy. To update this configuration,
  # update the conda-forge.yml and/or the recipe/meta.yaml.
  # -*- mode: yaml -*-
  
  jobs:
  - job: win
    pool:
      vmImage: windows-2022
    strategy:
      matrix:
        win_64_python2.7:
          CONFIG: win_64_python2.7
          UPLOAD_PACKAGES: 'True'
        win_64_python3.5:
          CONFIG: win_64_python3.5
          UPLOAD_PACKAGES: 'True'
    timeoutInMinutes: 360
    variables:
      CONDA_BLD_PATH: D:\\bld\\
      UPLOAD_TEMP: D:\\tmp
  
    steps:
      - script: |
          choco install pkg0 -fdv -y --debug
        displayName: "Install Chocolatey Package: pkg0"
  
      - script: |
          choco install pkg1 --version=X.Y.Z -fdv -y --debug
        displayName: "Install Chocolatey Package: pkg1 --version=X.Y.Z"
  
  
      - task: PythonScript@0
        displayName: 'Download Miniforge'
        inputs:
          scriptSource: inline
          script: |
            import urllib.request
            url = 'https://github.com/conda-forge/miniforge/releases/latest/download/Mambaforge-Windows-x86_64.exe'
            path = r"$(Build.ArtifactStagingDirectory)/Miniforge.exe"
            urllib.request.urlretrieve(url, path)
  
      - script: |
          start /wait "" %BUILD_ARTIFACTSTAGINGDIRECTORY%\Miniforge.exe /InstallationType=JustMe /RegisterPython=0 /S /D=C:\Miniforge
        displayName: Install Miniforge
  
      - powershell: Write-Host "##vso[task.prependpath]C:\Miniforge\Scripts"
        displayName: Add conda to PATH
  
      - script: |
          call ".scripts\run_win_build.bat"
        displayName: Run Windows build
        env:
          PYTHONUNBUFFERED: 1
          CONFIG: $(CONFIG)
          CI: azure
          flow_run_id: azure_$(Build.BuildNumber).$(System.JobAttempt)
          remote_url: $(Build.Repository.Uri)
          sha: $(Build.SourceVersion)
          UPLOAD_PACKAGES: $(UPLOAD_PACKAGES)
          UPLOAD_TEMP: $(UPLOAD_TEMP)
          UPLOAD_ON_BRANCH: foo-branch
          BINSTAR_TOKEN: $(BINSTAR_TOKEN)
  '''
# ---
# name: test_conda_build_tools[conda-build]
  '''
  WARNING  conda_smithy.configure_feedstock:configure_feedstock.py:2236 conda-forge.yml: $.build_with_mambabuild = True -> 'Build With Mambabuild' is deprecated.
  build_with_mambabuild is deprecated, use `conda_build_tool` instead.
  WARNING  conda_smithy.configure_feedstock:configure_feedstock.py:2236 conda-forge.yml: $.build_with_mambabuild = False -> 'Build With Mambabuild' is deprecated.
  build_with_mambabuild is deprecated, use `conda_build_tool` instead.
  WARNING  conda_smithy.configure_feedstock:configure_feedstock.py:2236 conda-forge.yml: $.conda_build_tool = does-not-exist -> 'does-not-exist' is not valid under any of the given schemas
  
  '''
# ---
# name: test_conda_build_tools[rattler-build]
  '''
  WARNING  conda_smithy.configure_feedstock:configure_feedstock.py:2236 conda-forge.yml: $.conda_build_tool = does-not-exist -> 'does-not-exist' is not valid under any of the given schemas
  
  '''
# ---
# name: test_get_used_key_values_by_input_order[squished_input_variants0-squished_used_variants0-all_used_vars0][get_used_key_values_by_input_order]
  dict({
    'c_compiler': list([
      'gcc',
    ]),
    'c_compiler_version': list([
      '10',
      '12',
    ]),
    'cdt_name': list([
      'cos7',
      'cos6',
    ]),
    'channel_sources': list([
      'conda-forge',
    ]),
    'channel_targets': list([
      'conda-forge main',
    ]),
    'cuda_compiler': list([
      'nvcc',
      'None',
    ]),
    'cuda_compiler_version': list([
      '11.2',
      'None',
    ]),
    'docker_image': list([
      'quay.io/condaforge/linux-anvil-cuda:11.2',
      'quay.io/condaforge/linux-anvil-cos7-x86_64',
    ]),
    'pin_run_as_build': dict({
      'flann': dict({
        'max_pin': 'x.x.x',
      }),
      'graphviz': dict({
        'max_pin': 'x',
      }),
      'libsvm': dict({
        'max_pin': 'x',
      }),
      'netcdf-cxx4': dict({
        'max_pin': 'x.x',
      }),
      'occt': dict({
        'max_pin': 'x.x',
      }),
      'poppler': dict({
        'max_pin': 'x.x',
      }),
      'python': dict({
        'max_pin': 'x.x',
        'min_pin': 'x.x',
      }),
      'r-base': dict({
        'max_pin': 'x.x',
        'min_pin': 'x.x',
      }),
      'vlfeat': dict({
        'max_pin': 'x.x.x',
      }),
    }),
    'target_platform': list([
      'linux-64',
    ]),
    'zip_keys': list([
      tuple(
        'arrow_cpp',
        'libarrow',
        'libarrow_all',
      ),
      tuple(
        'c_compiler_version',
        'cxx_compiler_version',
        'fortran_compiler_version',
        'cuda_compiler',
        'cuda_compiler_version',
        'cdt_name',
        'docker_image',
      ),
      tuple(
        'c_stdlib',
        'c_stdlib_version',
      ),
      tuple(
        'libgrpc',
        'libprotobuf',
      ),
      tuple(
        'python',
        'numpy',
        'python_impl',
      ),
    ]),
  })
# ---
# name: test_noarch_platforms_bad_yaml[conda-build]
  '''
  WARNING  conda_smithy.configure_feedstock:configure_feedstock.py:2236 conda-forge.yml: $.noarch_platforms = ['eniac', 'zx80'] -> ['eniac', 'zx80'] is not valid under any of the given schemas
  
  '''
# ---
# name: test_noarch_platforms_bad_yaml[rattler-build]
  '''
  WARNING  conda_smithy.configure_feedstock:configure_feedstock.py:2236 conda-forge.yml: $.noarch_platforms = ['eniac', 'zx80'] -> ['eniac', 'zx80'] is not valid under any of the given schemas
  
  '''
# ---
# name: test_stdlib_deployment_target[conda-build]
  '''
  MACOSX_DEPLOYMENT_TARGET:
  - '10.14'
  MACOSX_SDK_VERSION:
  - '10.14'
  c_compiler:
  - clang
  c_stdlib:
  - macosx_deployment_target
  c_stdlib_version:
  - '10.14'
  target_platform:
  - osx-64
  
  '''
# ---
# name: test_stdlib_on_azure[conda-build][linux content]
  '''
  c_compiler:
  - gcc
  c_stdlib:
  - sysroot
  c_stdlib_version:
  - '2.12'
  docker_image:
  - quay.io/condaforge/linux-anvil-comp7
  target_platform:
  - linux-64
  
  '''
# ---
# name: test_stdlib_on_azure[conda-build][osx content]
  '''
  MACOSX_DEPLOYMENT_TARGET:
  - '10.9'
  MACOSX_SDK_VERSION:
  - '10.9'
  c_compiler:
  - clang
  c_stdlib:
  - macosx_deployment_target
  c_stdlib_version:
  - '10.9'
  target_platform:
  - osx-64
  
  '''
# ---
# name: test_stdlib_on_azure[conda-build][win content]
  '''
  c_compiler:
  - vs2017
  c_stdlib:
  - vs
  target_platform:
  - win-64
  
  '''
# ---
# name: test_upload_on_branch_appveyor[conda-build][config]
  dict({
    'upload_on_branch': 'foo-branch',
  })
# ---
# name: test_upload_on_branch_appveyor[conda-build][deploy script]
  list([
    'set "GIT_BRANCH=%APPVEYOR_REPO_BRANCH%"',
    'set "FEEDSTOCK_NAME=%APPVEYOR_REPO_NAME:*/=%"',
    'set "UPLOAD_ON_BRANCH=foo-branch"',
    dict({
      'cmd': 'upload_package  .\\ .\\recipe .ci_support\\%CONFIG%.yaml',
    }),
  ])
# ---
# name: test_upload_on_branch_appveyor[rattler-build][config]
  dict({
    'upload_on_branch': 'foo-branch',
  })
# ---
# name: test_upload_on_branch_appveyor[rattler-build][deploy script]
  list([
    'set "GIT_BRANCH=%APPVEYOR_REPO_BRANCH%"',
    'set "FEEDSTOCK_NAME=%APPVEYOR_REPO_NAME:*/=%"',
    'set "UPLOAD_ON_BRANCH=foo-branch"',
    dict({
      'cmd': 'upload_package  .\\ .\\recipe .ci_support\\%CONFIG%.yaml',
    }),
  ])
# ---
# name: test_upload_on_branch_azure[conda-build][config]
  dict({
    'upload_on_branch': 'foo-branch',
  })
# ---
# name: test_upload_on_branch_azure[conda-build][content linux]
  '''
  export CI=azure
  export flow_run_id=azure_$(Build.BuildNumber).$(System.JobAttempt)
  export remote_url=$(Build.Repository.Uri)
  export sha=$(Build.SourceVersion)
  export GIT_BRANCH=$BUILD_SOURCEBRANCHNAME
  export FEEDSTOCK_NAME=$(basename ${BUILD_REPOSITORY_NAME})
  export UPLOAD_ON_BRANCH="foo-branch"
  if [[ "${BUILD_REASON:-}" == "PullRequest" ]]; then
    export IS_PR_BUILD="True"
  else
    export IS_PR_BUILD="False"
  fi
  .scripts/run_docker_build.sh
  
  '''
# ---
# name: test_upload_on_branch_azure[conda-build][content osx]
  '''
  export CI=azure
  export flow_run_id=azure_$(Build.BuildNumber).$(System.JobAttempt)
  export remote_url=$(Build.Repository.Uri)
  export sha=$(Build.SourceVersion)
  export OSX_FORCE_SDK_DOWNLOAD="1"
  export GIT_BRANCH=$BUILD_SOURCEBRANCHNAME
  export FEEDSTOCK_NAME=$(basename ${BUILD_REPOSITORY_NAME})
  export UPLOAD_ON_BRANCH="foo-branch"
  if [[ "${BUILD_REASON:-}" == "PullRequest" ]]; then
    export IS_PR_BUILD="True"
  else
    export IS_PR_BUILD="False"
  fi
  ./.scripts/run_osx_build.sh
  
  '''
# ---
# name: test_upload_on_branch_azure[rattler-build][config]
  dict({
    'upload_on_branch': 'foo-branch',
  })
# ---
# name: test_upload_on_branch_azure[rattler-build][content linux]
  '''
  export CI=azure
  export flow_run_id=azure_$(Build.BuildNumber).$(System.JobAttempt)
  export remote_url=$(Build.Repository.Uri)
  export sha=$(Build.SourceVersion)
  export GIT_BRANCH=$BUILD_SOURCEBRANCHNAME
  export FEEDSTOCK_NAME=$(basename ${BUILD_REPOSITORY_NAME})
  export UPLOAD_ON_BRANCH="foo-branch"
  if [[ "${BUILD_REASON:-}" == "PullRequest" ]]; then
    export IS_PR_BUILD="True"
  else
    export IS_PR_BUILD="False"
  fi
  .scripts/run_docker_build.sh
  
  '''
# ---
# name: test_upload_on_branch_azure[rattler-build][content osx]
  '''
  export CI=azure
  export flow_run_id=azure_$(Build.BuildNumber).$(System.JobAttempt)
  export remote_url=$(Build.Repository.Uri)
  export sha=$(Build.SourceVersion)
  export OSX_FORCE_SDK_DOWNLOAD="1"
  export GIT_BRANCH=$BUILD_SOURCEBRANCHNAME
  export FEEDSTOCK_NAME=$(basename ${BUILD_REPOSITORY_NAME})
  export UPLOAD_ON_BRANCH="foo-branch"
  if [[ "${BUILD_REASON:-}" == "PullRequest" ]]; then
    export IS_PR_BUILD="True"
  else
    export IS_PR_BUILD="False"
  fi
  ./.scripts/run_osx_build.sh
  
  '''
# ---
# name: test_webservices_action_exists[conda-build]
  dict({
    'jobs': dict({
      'webservices': dict({
        'name': 'webservices',
        'runs-on': 'ubuntu-latest',
        'steps': list([
          dict({
            'id': 'webservices',
            'name': 'webservices',
            'uses': 'conda-forge/webservices-dispatch-action@main',
            'with': dict({
              'github_token': '${{ secrets.GITHUB_TOKEN }}',
              'rerendering_github_token': '${{ secrets.RERENDERING_GITHUB_TOKEN }}',
            }),
          }),
        ]),
      }),
    }),
    True: 'repository_dispatch',
  })
# ---
# name: test_webservices_action_exists[rattler-build]
  dict({
    'jobs': dict({
      'webservices': dict({
        'name': 'webservices',
        'runs-on': 'ubuntu-latest',
        'steps': list([
          dict({
            'id': 'webservices',
            'name': 'webservices',
            'uses': 'conda-forge/webservices-dispatch-action@main',
            'with': dict({
              'github_token': '${{ secrets.GITHUB_TOKEN }}',
              'rerendering_github_token': '${{ secrets.RERENDERING_GITHUB_TOKEN }}',
            }),
          }),
        ]),
      }),
    }),
    True: 'repository_dispatch',
  })
# ---
